src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -0,0 +1,151 @@
+      int64_t failure_threshold = RayConfig::instance().health_check_failure_threshold());
+
+  ~GcsHealthCheckManager();
+
+  /// Start to track the healthiness of a node.
+  ///
+  /// \param node_id The id of the node.
+  /// \param channel The gRPC channel to the node.
+  void AddNode(const NodeID &node_id, std::shared_ptr<grpc::Channel> channel);
+
+  /// Stop tracking the healthiness of a node.
+  ///
+  /// \param node_id The id of the node to stop tracking.
+  void RemoveNode(const NodeID &node_id);
+
+  /// Return all the nodes monitored.
+  ///
+  /// \return A list of node id which are being monitored by this class.
+  std::vector<NodeID> GetAllNodes() const;
+
+ private:
+  void FailNode(const NodeID &node_id);
+
+  /// This is for testing. We'll use mock timer in gtest.
+#if defined(_TESTING_RAY_TIMER)
+  using Timer = _TESTING_RAY_TIMER;
+#else
+  using Timer = boost::asio::deadline_timer;
+#endif
+
+  /// The context for the health check. It's to support unary call.
+  /// It can be updated to support streaming call for efficiency.
+  class HealthCheckContext {
+   public:
+    HealthCheckContext(GcsHealthCheckManager *_manager,
+                       std::shared_ptr<grpc::Channel> channel,
+                       NodeID node_id)
+        : manager(_manager),
+          node_id(node_id),
+          timer(manager->io_service_),
+          health_check_remaining(manager->failure_threshold_) {
+      stub = grpc::health::v1::Health::NewStub(channel);
+      timer.expires_from_now(boost::posix_time::milliseconds(manager->initial_delay_ms_));
+      timer.async_wait([this](auto ec) {
+        if (ec != boost::asio::error::operation_aborted) {
+          StartHealthCheck();
+        }
+      });
+    }
+
+    ~HealthCheckContext() {
+      timer.cancel();
+      if (context != nullptr) {
+        context->TryCancel();
+      }
+    }
+
+   private:
+    void StartHealthCheck();
+
+    GcsHealthCheckManager *manager;
+
+    NodeID node_id;
+
+    /// gRPC related fields
+    std::unique_ptr<::grpc::health::v1::Health::Stub> stub;
+
+    // The context is used in the gRPC callback which is in another
+    // thread, so we need it to be a shared_ptr.
+    std::shared_ptr<grpc::ClientContext> context;
+    ::grpc::health::v1::HealthCheckRequest request;
+    ::grpc::health::v1::HealthCheckResponse response;
+
+    /// The timer is used to do async wait before the next try.
+    Timer timer;
+
+    /// The remaining check left. If it reaches 0, the node will be marked as dead.
+    int64_t health_check_remaining;
+  };
+
+  /// The main service. All method needs to run on this thread.
+  instrumented_io_context &io_service_;
+
+  /// Callback when the node failed.
+  std::function<void(const NodeID &)> on_node_death_callback_;
+
+  /// The context of the health check for each nodes.
+  absl::flat_hash_map<NodeID, std::unique_ptr<HealthCheckContext>>
+      inflight_health_checks_;
```

Comment:
would `health_checked_nodes_` or simply `nodes_` be a better name?

Reply:
maybe `health_check_contexts_`?